scripts/check_docs_formatted.py

```diff
@@ -0,0 +1,195 @@
+            " rewritten to:",
+        )
+
+        # Add indentation so that snipped can be copied directly to docs
+        for line in new_contents.splitlines():
+            output_line = "///"
+            if len(line) > 0:
+                output_line = f"{output_line} {line}"
+
+            print(output_line)
+
+        print("\n")
+
+        return 1
+
+    return 0
+
+
+def main(argv: Sequence[str] | None = None) -> int:
+    """Check code snippets in docs are formatted by black."""
+    parser = argparse.ArgumentParser(
+        description="Check code snippets in docs are formatted by black.",
+    )
+    parser.add_argument("--skip-errors", action="store_true")
+    parser.add_argument("--generate-docs", action="store_true")
+    args = parser.parse_args(argv)
+
+    if args.generate_docs:
+        # Generate docs
+        from generate_mkdocs import main as generate_docs
+
+        generate_docs()
+
+    # Get static docs
+    static_docs = []
+    for file in os.listdir("docs"):
+        if file.endswith(".md"):
+            static_docs.append(Path("docs") / file)
+
+    # Check rules generated
+    if not Path("docs/rules").exists():
+        print("Please generate rules first.")
+        return 1
+
+    # Get generated rules
+    generated_docs = []
+    for file in os.listdir("docs/rules"):
+        if file.endswith(".md"):
+            generated_docs.append(Path("docs/rules") / file)
+
+    if len(generated_docs) == 0:
+        print("Please generate rules first.")
+        return 1
+
+    black_mode = Mode(
+        target_versions={TargetVersion[val.upper()] for val in TARGET_VERSIONS},
+    )
+
+    # For some docs, we don't want black to fix the formatting as this would remove the
+    # reason for the example. These rules will be stored in
+    # `scripts/known_formatting_violations.txt`
+
+    with Path("scripts/known_rule_formatting_violations.txt").open() as f:
+        known_formatting_violations = f.read().splitlines()
+
+    # For some docs, black is unable to parse the example code. These rules will be
+    # stored at `scripts/known_rule_parse_errors.txt`
+
+    with Path("scripts/known_rule_parse_errors.txt").open() as f:
+        known_parse_errors = f.read().splitlines()
+
+    # Check known formatting violations and parse errors are sorted alphabetically and
+    # have no duplicates. This will reduce the diff when adding new violations
+
+    for known_list, file_string in [
+        (known_formatting_violations, "formatting violations"),
+        (known_parse_errors, "parse errors"),
+    ]:
+        if known_list != sorted(known_list):
+            print(
+                f"Known {file_string} is not sorted alphabetically. Please sort and"
+                " re-run.",
+            )
+            return 1
+
+        if len(known_list) != len(set(known_list)):
+            print(f"Known {file_string} has duplicates. Please remove them and re-run.")
+            return 1
```

Comment:
Nit: It would be nice if the script could print the name of the duplicated entry.

Reply:
Updated in 673035f82328d4522253e92c0b3652017aa1cbf0